File: src/test/java/io/georocket/commands/DeleteCommandTest.java
package io.georocket.commands;

import static com.github.tomakehurst.wiremock.client.WireMock.aResponse;
import static com.github.tomakehurst.wiremock.client.WireMock.delete;
import static com.github.tomakehurst.wiremock.client.WireMock.deleteRequestedFor;
import static com.github.tomakehurst.wiremock.client.WireMock.stubFor;
import static com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo;
import static com.github.tomakehurst.wiremock.client.WireMock.verify;

import org.junit.Test;
import org.junit.runner.RunWith;

import com.github.tomakehurst.wiremock.client.VerificationException;

import io.vertx.ext.unit.Async;
import io.vertx.ext.unit.TestContext;
import io.vertx.ext.unit.junit.VertxUnitRunner;

/**
 * Test for {@link DeleteCommand}
 * @author Michel Kraemer
 */
@RunWith(VertxUnitRunner.class)
public class DeleteCommandTest extends CommandTestBase<DeleteCommand> {
  @Override
  protected DeleteCommand createCommand() {
    return new DeleteCommand();
  }
  
  /**
   * Test no layer
   * @param context the test context
   * @throws Exception if something goes wrong
   */
  @Test
  public void noLayer(TestContext context) throws Exception {
    context.assertEquals(1, cmd.run(new String[] { }, in, out));
  }
  
  /**
   * Test empty layer
   * @param context the test context
   * @throws Exception if something goes wrong
   */
  @Test
  public void emptyLayer(TestContext context) throws Exception {
    context.assertEquals(1, cmd.run(new String[] { "" }, in, out));
  }
  
  /**
   * Verify that a certain DELETE request has been made
   * @param url the request URL
   * @param context the current test context
   */
  protected void verifyDeleted(String url, TestContext context) {
    try {
      verify(deleteRequestedFor(urlEqualTo(url)));
    } catch (VerificationException e) {
      context.fail(e);
    }
  }
  
  /**
   * Test a delete with a simple query
   * @param context the test context
   * @throws Exception if something goes wrong
   */
  @Test
  public void simpleQueryDelete(TestContext context) throws Exception {
    String url = "/store/?search=test";
    stubFor(delete(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(204)));
    
    Async async = context.async();
    cmd.setEndHandler(exitCode -> {
      context.assertEquals(0, exitCode);
      verifyDeleted(url, context);
      async.complete();
    });
    
    cmd.run(new String[] { "test" }, in, out);
  }

  /**
   * Test a delete with query that consists of two terms
   * @param context the test context
   * @throws Exception if something goes wrong
   */
  @Test
  public void twoTermsQueryDelete(TestContext context) throws Exception {
    String url = "/store/?search=test1+test2";
    stubFor(delete(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(204)));
    
    Async async = context.async();
    cmd.setEndHandler(exitCode -> {
      context.assertEquals(0, exitCode);
      verifyDeleted(url, context);
      async.complete();
    });
    
    cmd.run(new String[] { "test1", "test2" }, in, out);
  }
  
  /**
   * Test to delete the root layer
   * @param context the test context
   * @throws Exception if something goes wrong
   */
  @Test
  public void rootLayer(TestContext context) throws Exception {
    String url = "/store/";
    stubFor(delete(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(204)));
    
    Async async = context.async();
    cmd.setEndHandler(exitCode -> {
      context.assertEquals(0, exitCode);
      verifyDeleted(url, context);
      async.complete();
    });
    
    cmd.run(new String[] { "-l", "/" }, in, out);
  }
  
  /**
   * Test a delete with a layer but no query
   * @param context the test context
   * @throws Exception if something goes wrong
   */
  @Test
  public void layerNoQuery(TestContext context) throws Exception {
    String url = "/store/hello/world/";
    stubFor(delete(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(204)));
    
    Async async = context.async();
    cmd.setEndHandler(exitCode -> {
      context.assertEquals(0, exitCode);
      verifyDeleted(url, context);
      async.complete();
    });
    
    cmd.run(new String[] { "-l", "hello/world" }, in, out);
  }
  
  /**
   * Test a delete with a layer and a query
   * @param context the test context
   * @throws Exception if something goes wrong
   */
  @Test
  public void layerQuery(TestContext context) throws Exception {
    String url = "/store/hello/world/?search=test";
    stubFor(delete(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(204)));
    
    Async async = context.async();
    cmd.setEndHandler(exitCode -> {
      context.assertEquals(0, exitCode);
      verifyDeleted(url, context);
      async.complete();
    });
    
    cmd.run(new String[] { "-l", "hello/world", "test" }, in, out);
  }
}

File: georocket-cli/src/test/kotlin/io/georocket/commands/DeleteCommandTest.kt
package io.georocket.commands

import com.github.tomakehurst.wiremock.client.VerificationException
import com.github.tomakehurst.wiremock.client.WireMock.aResponse
import com.github.tomakehurst.wiremock.client.WireMock.delete
import com.github.tomakehurst.wiremock.client.WireMock.deleteRequestedFor
import com.github.tomakehurst.wiremock.client.WireMock.stubFor
import com.github.tomakehurst.wiremock.client.WireMock.urlEqualTo
import com.github.tomakehurst.wiremock.client.WireMock.verify
import io.vertx.ext.unit.TestContext
import io.vertx.ext.unit.junit.VertxUnitRunner
import io.vertx.kotlin.coroutines.dispatcher
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Test for [DeleteCommand]
 */
@RunWith(VertxUnitRunner::class)
class DeleteCommandTest : CommandTestBase<DeleteCommand>() {
  override val cmd = DeleteCommand()

  /**
   * Test no layer
   */
  @Test
  fun noLayer(context: TestContext) {
    val async = context.async()
    GlobalScope.launch(rule.vertx().dispatcher()) {
      val exitCode = cmd.runAwait(arrayOf(), input, out)
      context.assertEquals(1, exitCode)
      async.complete()
    }
  }

  /**
   * Test empty layer
   */
  @Test
  fun emptyLayer(context: TestContext) {
    val async = context.async()
    GlobalScope.launch(rule.vertx().dispatcher()) {
      val exitCode = cmd.runAwait(arrayOf(""), input, out)
      context.assertEquals(1, exitCode)
      async.complete()
    }
  }

  /**
   * Verify that a certain DELETE request has been made
   */
  private fun verifyDeleted(url: String, context: TestContext) {
    try {
      verify(deleteRequestedFor(urlEqualTo(url)))
    } catch (e: VerificationException) {
      context.fail(e)
    }
  }

  /**
   * Test a delete with a simple query
   */
  @Test
  fun simpleQueryDelete(context: TestContext) {
    val url = "/store/?search=test"
    stubFor(delete(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(204)))

    val async = context.async()
    GlobalScope.launch(rule.vertx().dispatcher()) {
      val exitCode = cmd.runAwait(arrayOf("test"), input, out)
      context.assertEquals(0, exitCode)
      verifyDeleted(url, context)
      async.complete()
    }
  }

  /**
   * Test a delete with query that consists of two terms
   */
  @Test
  fun twoTermsQueryDelete(context: TestContext) {
    val url = "/store/?search=test1%20test2"
    stubFor(delete(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(204)))

    val async = context.async()
    GlobalScope.launch(rule.vertx().dispatcher()) {
      val exitCode = cmd.runAwait(arrayOf("test1", "test2"), input, out)
      context.assertEquals(0, exitCode)
      verifyDeleted(url, context)
      async.complete()
    }
  }

  /**
   * Test to delete the root layer
   */
  @Test
  fun rootLayer(context: TestContext) {
    val url = "/store/"
    stubFor(delete(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(204)))

    val async = context.async()
    GlobalScope.launch(rule.vertx().dispatcher()) {
      val exitCode = cmd.runAwait(arrayOf("-l", "/"), input, out)
      context.assertEquals(0, exitCode)
      verifyDeleted(url, context)
      async.complete()
    }
  }

  /**
   * Test a delete with a layer but no query
   */
  @Test
  fun layerNoQuery(context: TestContext) {
    val url = "/store/hello/world/"
    stubFor(delete(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(204)))

    val async = context.async()
    GlobalScope.launch(rule.vertx().dispatcher()) {
      val exitCode = cmd.runAwait(arrayOf("-l", "hello/world"), input, out)
      context.assertEquals(0, exitCode)
      verifyDeleted(url, context)
      async.complete()
    }
  }

  /**
   * Test a delete with a layer and a query
   */
  @Test
  fun layerQuery(context: TestContext) {
    val url = "/store/hello/world/?search=test"
    stubFor(delete(urlEqualTo(url))
        .willReturn(aResponse()
            .withStatus(204)))

    val async = context.async()
    GlobalScope.launch(rule.vertx().dispatcher()) {
      val exitCode = cmd.runAwait(arrayOf("-l", "hello/world", "test"), input, out)
      context.assertEquals(0, exitCode)
      verifyDeleted(url, context)
      async.complete()
    }
  }
}
